File: server/app/adapters/mimo_tts.py
from __future__ import annotations

import base64
from dataclasses import dataclass

import httpx

from app.config import AppConfig
from app.models import TtsStatus


def _clip(value: str, max_length: int) -> str:
    return value if len(value) <= max_length else value[:max_length] + "..."


def _join_url(base_url: str, path: str) -> str:
    return f"{base_url.rstrip('/')}/{path.lstrip('/')}"


@dataclass(frozen=True)
class SynthesizedAudio:
    buffer: bytes
    provider: str
    format: str
    mime_type: str


class MimoTtsAdapter:
    REQUEST_TIMEOUT_SEC = 12

    def __init__(self, config: AppConfig) -> None:
        self.config = config

    def get_status(self) -> TtsStatus:
        configured = bool(self.config.mimo_api_key and self.config.mimo_tts_voice)
        return TtsStatus(
            configured=configured,
            provider="mimo" if configured else "tts-disabled",
            format=self.config.mimo_tts_format,
            voice_configured=bool(self.config.mimo_tts_voice),
            detail=(
                f"Mimo chat-completions TTS is ready via proxy {self.config.mimo_proxy_url}."
                if configured and self.config.mimo_proxy_url
                else "Mimo chat-completions TTS is ready."
                if configured
                else "Missing MIMO_API_KEY or MIMO_TTS_VOICE. Narration audio is disabled."
            ),
        )

    async def synthesize(self, text: str) -> SynthesizedAudio:
        if not self.config.mimo_api_key or not self.config.mimo_tts_voice:
            raise RuntimeError("Mimo TTS is not configured.")

        async with httpx.AsyncClient(
            timeout=self.REQUEST_TIMEOUT_SEC,
            proxy=self.config.mimo_proxy_url,
        ) as client:
            response = await client.post(
                _join_url(self.config.mimo_base_url, "/chat/completions"),
                headers={
                    "Content-Type": "application/json",
                    "api-key": self.config.mimo_api_key,
                },
                json={
                    "model": self.config.mimo_tts_model,
                    "messages": [
                        {
                            "role": "user",
                            "content": "请用清晰自然的普通话逐字朗读下一条 assistant 消息。不要改写，不要解释，不要加入任何额外内容。",
                        },
                        {"role": "assistant", "content": text},
                    ],
                    "audio": {
                        "format": self.config.mimo_tts_format,
                        "voice": self.config.mimo_tts_voice,
                    },
                },
            )

        if response.status_code < 200 or response.status_code >= 300:
            raise RuntimeError(f"Mimo request failed ({response.status_code}): {_clip(response.text, 200)}")

        payload = response.json()
        audio = (
            ((payload.get("choices") or [{}])[0].get("message") or {})
            .get("audio")
            or {}
        )
        base64_audio = audio.get("data")
        if not isinstance(base64_audio, str) or not base64_audio.strip():
            raise RuntimeError("Mimo returned JSON without choices[0].message.audio.data.")

        return SynthesizedAudio(
            buffer=base64.b64decode(base64_audio),
            provider="mimo",
            format=self.config.mimo_tts_format,
            mime_type=self._mime_type_for_format(self.config.mimo_tts_format),
        )

    def _mime_type_for_format(self, fmt: str) -> str:
        if fmt == "wav":
            return "audio/wav"
        if fmt in {"ogg", "opus"}:
            return "audio/ogg"
        if fmt == "flac":
            return "audio/flac"
        if fmt == "aac":
            return "audio/aac"
        return "audio/mpeg"

File: .indio-backups/server-ts-20260507-002053/src/adapters/mimo-tts.ts
import type { AppConfig } from "../config.js";
import type { TtsStatus } from "@indio/contracts";
import { ProxyAgent, fetch as undiciFetch } from "undici";

interface SynthesizedAudio {
  buffer: Buffer;
  provider: string;
  format: string;
  mimeType: string;
}

function clip(value: string, maxLength: number): string {
  return value.length <= maxLength ? value : `${value.slice(0, maxLength)}...`;
}

function asObject(value: unknown): Record<string, unknown> | null {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    return value as Record<string, unknown>;
  }

  return null;
}

function asString(value: unknown): string | null {
  return typeof value === "string" && value.trim() ? value : null;
}

function asArray(value: unknown): unknown[] | null {
  return Array.isArray(value) ? value : null;
}

function joinUrl(baseUrl: string, path: string): string {
  return new URL(path.replace(/^\//, ""), baseUrl.endsWith("/") ? baseUrl : `${baseUrl}/`).toString();
}

export class MimoTtsAdapter {
  private static readonly REQUEST_TIMEOUT_MS = 12_000;
  private readonly proxyAgent: ProxyAgent | null;

  constructor(private readonly config: AppConfig) {
    this.proxyAgent = config.mimoProxyUrl ? new ProxyAgent(config.mimoProxyUrl) : null;
  }

  getStatus(): TtsStatus {
    const configured = Boolean(this.config.mimoApiKey && this.config.mimoTtsVoice);

    return {
      configured,
      provider: configured ? "mimo" : "tts-disabled",
      format: this.config.mimoTtsFormat,
      voiceConfigured: Boolean(this.config.mimoTtsVoice),
      detail: configured
        ? this.config.mimoProxyUrl
          ? `Mimo chat-completions TTS is ready via proxy ${this.config.mimoProxyUrl}.`
          : "Mimo chat-completions TTS is ready."
        : "Missing MIMO_API_KEY or MIMO_TTS_VOICE. Narration audio is disabled."
    };
  }

  async synthesize(text: string): Promise<SynthesizedAudio> {
    if (!this.config.mimoApiKey || !this.config.mimoTtsVoice) {
      throw new Error("Mimo TTS is not configured.");
    }

    let response: Awaited<ReturnType<typeof undiciFetch>>;

    try {
      response = await undiciFetch(joinUrl(this.config.mimoBaseUrl, "/chat/completions"), {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "api-key": this.config.mimoApiKey
        },
        body: JSON.stringify({
          model: this.config.mimoTtsModel,
          messages: [
            {
              role: "user",
              content: "请用清晰自然的普通话逐字朗读下一条 assistant 消息。不要改写，不要解释，不要加入任何额外内容。"
            },
            {
              role: "assistant",
              content: text
            }
          ],
          audio: {
            format: this.config.mimoTtsFormat,
            voice: this.config.mimoTtsVoice
          }
        }),
        signal: AbortSignal.timeout(MimoTtsAdapter.REQUEST_TIMEOUT_MS),
        ...(this.proxyAgent ? { dispatcher: this.proxyAgent } : {})
      });
    } catch (error) {
      if (error instanceof Error && error.name === "TimeoutError") {
        throw new Error(`Mimo request timed out after ${MimoTtsAdapter.REQUEST_TIMEOUT_MS}ms.`);
      }

      throw error;
    }

    if (!response.ok) {
      throw new Error(`Mimo request failed (${response.status}): ${clip(await response.text(), 200)}`);
    }

    const contentType = response.headers.get("content-type") ?? "application/octet-stream";

    if (!contentType.includes("application/json")) {
      throw new Error(`Unexpected Mimo content-type: ${contentType}`);
    }

    const payload = asObject(await response.json());
    const choices = asArray(payload?.choices);
    const firstChoice = asObject(choices?.[0]);
    const message = asObject(firstChoice?.message);
    const audio = asObject(message?.audio);
    const base64Audio = asString(audio?.data);

    if (!base64Audio) {
      throw new Error("Mimo returned JSON without choices[0].message.audio.data.");
    }

    return {
      buffer: Buffer.from(base64Audio, "base64"),
      provider: "mimo",
      format: this.config.mimoTtsFormat,
      mimeType: this.mimeTypeForFormat(this.config.mimoTtsFormat)
    };
  }

  private mimeTypeForFormat(format: string): string {
    if (format === "wav") {
      return "audio/wav";
    }

    if (format === "ogg" || format === "opus") {
      return "audio/ogg";
    }

    if (format === "flac") {
      return "audio/flac";
    }

    if (format === "aac") {
      return "audio/aac";
    }

    return "audio/mpeg";
  }
}
